File: controllers/authController.js
const User = require('../models/User');

const authController = {
  // GET /login
  loginPage(req, res) {
    res.render('login', { user: req.session.user || null, error: null });
  },

  // POST /login
  login(req, res) {
    const { email, password } = req.body;
    const user = User.findByEmail(email);

    if (!user || !User.verifyPassword(password, user.password)) {
      return res.render('login', { user: null, error: 'Invalid email or password' });
    }

    req.session.user = {
      user_id: user.user_id,
      name: user.name,
      email: user.email,
      role: user.role
    };

    res.redirect('/dashboard');
  },

  // GET /register
  registerPage(req, res) {
    res.render('register', { user: req.session.user || null, error: null });
  },

  // POST /register
  register(req, res) {
    const { name, email, password, role } = req.body;

    try {
      const existing = User.findByEmail(email);
      if (existing) {
        return res.render('register', { user: null, error: 'Email already registered' });
      }

      const allowedRoles = ['customer', 'agent'];
      const userRole = allowedRoles.includes(role) ? role : 'customer';

      User.create(name, email, password, userRole);
      res.redirect('/login');
    } catch (err) {
      res.render('register', { user: null, error: 'Registration failed. Try again.' });
    }
  },

  // GET /logout
  logout(req, res) {
    req.session.destroy();
    res.redirect('/login');
  }
};

module.exports = authController;

File: src/main/java/com/example/insureflow/controller/AuthController.java
package com.example.insureflow.controller;

import com.example.insureflow.model.User;
import com.example.insureflow.model.UserRole;
import com.example.insureflow.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;
import jakarta.servlet.http.HttpSession;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;

@Controller
public class AuthController {
    @Autowired
    private UserService userService;

    @GetMapping("/login")
    public String loginPage(Model model, HttpSession session) {
        if (session.getAttribute("user") != null) {
            return "redirect:/dashboard";
        }
        return "login";
    }

    @PostMapping("/login")
    public String login(@RequestParam String email, @RequestParam String password, 
                       Model model, HttpSession session) {
        Optional<User> user = userService.authenticate(email, password);
        if (user.isPresent()) {
            // Set session attributes
            session.setAttribute("user", user.get());
            session.setAttribute("userId", user.get().getUserId());
            
            // Register with Spring Security
            Collection<SimpleGrantedAuthority> authorities = new ArrayList<>();
            authorities.add(new SimpleGrantedAuthority("ROLE_" + user.get().getRole().toString()));
            
            Authentication authentication = new UsernamePasswordAuthenticationToken(
                email, password, authorities
            );
            
            SecurityContext securityContext = SecurityContextHolder.createEmptyContext();
            securityContext.setAuthentication(authentication);
            SecurityContextHolder.setContext(securityContext);
            
            // Store security context in session
            session.setAttribute("SPRING_SECURITY_CONTEXT", securityContext);
            
            return "redirect:/dashboard";
        }
        model.addAttribute("error", "Invalid email or password");
        return "login";
    }

    @GetMapping("/register")
    public String registerPage(Model model, HttpSession session) {
        if (session.getAttribute("user") != null) {
            return "redirect:/dashboard";
        }
        return "register";
    }

    @PostMapping("/register")
    public String register(@RequestParam String name, @RequestParam String email, 
                          @RequestParam String password, @RequestParam(defaultValue = "customer") String role,
                          Model model, HttpSession session) {
        try {
            UserRole userRole = UserRole.fromLabel(role);
            userService.registerUser(name, email, password, userRole);
            return "redirect:/login";
        } catch (RuntimeException e) {
            model.addAttribute("error", e.getMessage());
            return "register";
        }
    }

    @GetMapping("/logout")
    public String logout(HttpSession session) {
        session.invalidate();
        return "redirect:/login";
    }

    @GetMapping("/")
    public String home(HttpSession session) {
        if (session.getAttribute("user") != null) {
            return "redirect:/dashboard";
        }
        return "redirect:/login";
    }
}
